contracts/farm/LiquidityMiningV1.sol

```diff
@@ -0,0 +1,708 @@
+		}
+
+		_updateRewardDebt(pool, user);
+		emit Withdraw(userAddress, _poolToken, _amount);
+	}
+
+	function _getUserAddress(address _user) internal view returns (address) {
+		address userAddress = msg.sender;
+		if (_user != address(0)) {
+			//only wrapper can pass _user parameter
+			require(msg.sender == wrapper || poolIdList[msg.sender] != 0, "only wrapper or pools may withdraw for a user");
+			userAddress = _user;
+		}
+		return userAddress;
+	}
+
+	function _updateReward(PoolInfo storage pool, UserInfo storage user) internal {
+		//update user accumulated reward
+		if (user.amount > 0) {
+			//add reward for the previous amount of deposited tokens
+			uint256 accumulatedReward = user.amount.mul(pool.accumulatedRewardPerShare).div(PRECISION).sub(user.rewardDebt);
+			user.accumulatedReward = user.accumulatedReward.add(accumulatedReward);
+		}
+	}
+
+	function _updateRewardDebt(PoolInfo storage pool, UserInfo storage user) internal {
+		//reward accumulated before amount update (should be subtracted during next reward calculation)
+		user.rewardDebt = user.amount.mul(pool.accumulatedRewardPerShare).div(PRECISION);
+	}
+
+	/**
+	 * @notice Send reward in SOV to the lockedSOV vault.
+	 * @param _user The user info, to get its reward share.
+	 * @param _userAddress The address of the user, to send SOV in its behalf.
+	 * @param _isStakingTokens The flag whether we need to stake tokens
+	 * @param _isCheckingBalance The flag whether we need to throw error or don't process reward if SOV balance isn't enough
+	 */
+	function _transferReward(
+		address _poolToken,
+		UserInfo storage _user,
+		address _userAddress,
+		bool _isStakingTokens,
+		bool _isCheckingBalance
+	) internal {
+		uint256 userAccumulatedReward = _user.accumulatedReward;
+
+		/// @dev Transfer if enough SOV balance on this LM contract.
+		uint256 balance = SOV.balanceOf(address(this));
+		if (balance >= userAccumulatedReward) {
+			totalUsersBalance = totalUsersBalance.sub(userAccumulatedReward);
+			_user.accumulatedReward = 0;
+
+			/// @dev Instead of transferring the reward to the LP (user),
+			///   deposit it into lockedSOV vault contract, but first
+			///   SOV deposit must be approved to move the SOV tokens
+			///   from this LM contract into the lockedSOV vault.
+			require(SOV.approve(address(lockedSOV), userAccumulatedReward), "Approve failed");
+			lockedSOV.deposit(_userAddress, userAccumulatedReward, unlockedImmediatelyPercent);
+
+			if (_isStakingTokens) {
+				lockedSOV.withdrawAndStakeTokensFrom(_userAddress);
+			}
+
+			/// @dev Event log.
+			emit RewardClaimed(_userAddress, _poolToken, userAccumulatedReward);
+		} else {
+			require(!_isCheckingBalance, "Claiming reward failed");
+		}
+	}
+
+	/**
+	 * @notice withdraws pool tokens without transferring reward tokens
+	 * @param _poolToken the address of pool token
+	 * @dev EMERGENCY ONLY
+	 */
+	function emergencyWithdraw(address _poolToken) external {
```

Comment:
this method should be "onlyBeforeMigrationGracePeriodFinished" because it can be called by any user